go-controller/pkg/ovnwebhook/nodeadmission.go

```diff
@@ -0,0 +1,191 @@
+		// it is allowed for the annotation to be set to <nodeName>
+		if (v.action == added || v.action == changed) && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s, it cannot be removed", util.OvnNodeMigratedZoneName, nodeName)
+	},
+}
+
+// hybridOverlayNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users hybrid overlay environments
+var hybridOverlayNodeAnnotationChecks = map[string]checkNodeAnnot{
+	hotypes.HybridOverlayDRMAC: nil,
+	hotypes.HybridOverlayDRIP:  nil,
+}
+
+type NodeAdmission struct {
+	annotationChecks  map[string]checkNodeAnnot
+	annotationKeys    sets.Set[string]
+	extraAllowedUsers sets.Set[string]
+}
+
+func NewNodeAdmissionWebhook(enableInterconnect, enableHybridOverlay bool, extraAllowedUsers ...string) *NodeAdmission {
+	checks := make(map[string]checkNodeAnnot)
+	maps.Copy(checks, commonNodeAnnotationChecks)
+	if enableInterconnect {
+		maps.Copy(checks, interconnectNodeAnnotationChecks)
+	}
+	if enableHybridOverlay {
+		maps.Copy(checks, hybridOverlayNodeAnnotationChecks)
+	}
+	return &NodeAdmission{
+		annotationChecks:  checks,
+		annotationKeys:    sets.New[string](maps.Keys(checks)...),
+		extraAllowedUsers: sets.New[string](extraAllowedUsers...),
+	}
+}
+
+var _ admission.CustomValidator = &NodeAdmission{}
+
+func (p NodeAdmission) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore deletion, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	nodeName, isOVNKubeNode := ovnkubeNodeIdentity(req.UserInfo)
+
+	changes := mapDiff(oldNode.Annotations, newNode.Annotations)
+	changedKeys := maps.Keys(changes)
+
+	if !isOVNKubeNode {
+		if !p.annotationKeys.HasAny(changedKeys...) {
+			// the user is not an ovnkube-node and hasn't changed any ovnkube-node annotations
+			return nil, nil
+		}
+
+		if !p.extraAllowedUsers.Has(req.UserInfo.Username) {
+			// the user is not in extraAllowedUsers
+			return nil, fmt.Errorf("user %q is not allowed to set the following annotations on node: %q: %v",
+				req.UserInfo.Username,
+				newNode.Name,
+				p.annotationKeys.Intersection(sets.New[string](changedKeys...)).UnsortedList())
+		}
+
+		// The user is not ovnkube-node, in this case the nodeName comes from the object
+		nodeName = newNode.Name
+	}
+
+	for _, key := range changedKeys {
```

Comment:
during upgrades, when we annotate that remote-migrated annotation, do we expect scale issues or disruptions because of this extra check?